### PR TITLE
Update SingletonSymNode to allow more comparisons

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108316
* __->__ #108315

In this PR:
- {in,}equality between singleton and plain ints returns false instead of erroring
- Morally define the semantic of j0 > c to be as if j0 represented an array [s_0, s_1, ... s_n] and s_k > c for all k
- Just like for equality, we don't actually want to do the comparison one by one, instead j0 is constrained to some range [min, max]. By default this range is [2, int64_t::max] so that it acts like a size and passes 0/1 specialization checks.
- In the future, we can define some API to allow users to constrain the range of their singletons

### CODE REVIEW

File: c10/core/ConstantSymNodeImpl.cpp [hunk 1]
```diff
@@ -1,23 +1,30 @@
...
+// singleton promotes the int into a constant symnode. If we'd like to
+// support more combinations in the future, we may need to implement some
+// kind of multiple dispatch.
+#define DEFINE_BINARY_OP(OP)                                           \
+  template <typename T>                                                \
+  c10::SymNode ConstantSymNodeImpl<T>::OP(const c10::SymNode& other) { \
+    TORCH_INTERNAL_ASSERT(other->singleton_int().has_value());         \
+    c10::raw::intrusive_ptr::incref(this);                             \
+    return other->OP(                                                  \
+        c10::intrusive_ptr<ConstantSymNodeImpl<T>>::reclaim(this));    \
```
Comment (ezyang) [inline c10/core/ConstantSymNodeImpl.cpp:17]: This is old code, but it's a bit weird to me that you incref and then reclaim, I feel like there should be a more direct API for this.

**Review by ezyang — APPROVED** [review]

**Comment by soulitzer** [comment]
Updates:
- realized I was not handling the self->op(other) --> other->inverse_op(self) case properly, whoops! (now we have to support all 4 inequalities)
- incref + reclaim = reclaim_copy

**Pull request closed without merging**